Adding the detailed triage process (#22393)
dotnet/aspnetcore: docs/TriageProcess.md

Managing a popular GitHub repo with a small team is not an easy task. It requires a good balance between creating new features while handling many investigations and bug fixes associated with existing ones.

> During the last couple of years the amount of incoming issues has been constanly growing. While this is a sign of a healthy framework and ecosystem surrounding it, it's becoming harder to react to all those issues.

Reviewer: ```suggestion
During the last couple of years the amount of incoming issues has been constantly growing. While this is a sign of a healthy framework and ecosystem surrounding it, it's becoming harder to react to all those issues.
```

> To be able to keep up with ever-evolving expectations, we're introducing a set of rules to help us better handling the incoming issues going forward.

Reviewer: ```suggestion
To be able to keep up with ever-evolving expectations, we're introducing a set of rules to help us better handle the incoming issues going forward.
```

## Goals
The goals of these rules are listed below in priority order:
- Make it easy to make triage decisions for each issue filed in this repository
- Be able to easily prioritize issues for each milestone
- Set proper expectations with customers regarding how issues are going to be handled

## Triage Process
The feature teams should be able look through every single issue filed against this repository and be able to make a quick call regarding the nature of the issue.
We will first categorize the issues and further handle these depending on the category the issue is in. The subsections below reprsent these categories and the rules we apply for them during our triage meeting.

> ### Feature requests

Author: We do reserve a lot of time for many deliverables, but we do leave time for stuff to come in. Based on this plan, we prioritize only a milestone ahead - so reasonable incoming feature requests are being evaluated each milestone and have potential to make into the next milestone.

> ### Feature requests

Reviewer 2: @Pilchie has that been your experience? When doing release planning with Andrew we costed out the major features for the release and cut what we didn't expect to fit. Even then we anticipate needing to cut more as we get closer to release. There's rarely room to add more significant features without cutting something else from the schedule. A few smaller ones might make it in.

We should be setting a clear expectation that most feature requests have to compete with the existing plan to make it in, and many of them won't make it in this release.

> ### Feature requests

Reviewer: I think we want to balance - we definitely pick "big rocks" for the release, and ensure that those fit.  Assuming there is any wiggle room after those, we fit in smaller things that don't require as much collaboration with other teams between them.  For those smaller things, I think it makes a lot of re-sense to re-check priorities on a regular basis.  Our plans don't have to be 100% locked for a year at a time.  With that said, I think there could be some wording here about the fact that many features *won't* make it into the current release.

> ### Feature requests

Author: I think this is already addressed. We do explicitly say, that feature-requests can be moved to `Backlog`. I don't think we should be more explicit than that.

> ### Feature requests

Author: That said, I would prefer to leave this as is.

As soon as we identify an issue represents an ask for a new feature, we will label the issue with the `enhancement` label.

> Most of the time, we will automatically move these issues into the `Next Sprint Planning` milestone for further review during the [next sprint planning meeting]().

Reviewer 2: Where are these broken links intended to refer to?

If we think the feature request is not aligned with our goals, we may choose to close it immediately.

> In some situations, however, we may choose to collect more feedback before acting on the issue. In these situations we will move the issue in the `Backlog` so that we can review it during our [release planning meeting]().

Reviewer 2: Are you planning to write a separate doc on release planning?

Author: Thanks for the reminder. These are supposed to refer to the `Milestone planning process` below. Will update that now.

If the impact is unclear or the is a very corner case scenario, we may move it to the `Backlog` milestone to further evaluate the impact by reviewing customer upvotes / comments at a later time.

### Investigations
In many situations it's not immediately clear whether a specific issue repoted is a bug or not. To be certain, the team will need to spend time to investigate it before making a call regarding the faith of the issue. In these situations we will apply the `investigate` label to the issue.
In many situations, these issues turn out to be a result of some kind of misconfiguration in the user code.
In some rare situations, however, these turn out to be caused by very impactful issues. So we will make a call during the triage whether we need to immediately investigate certain issues or not.
If not, we will move the investigation to the `Next Sprint Planning` to review during the [next sprint planning meeting]().

### Documentation requests
Some issues turn out to indicate user confusion around how to configure different aspects of the framework.

> When we determine such isssues, which impact many customers, we will mark these with the `Docs` label and move them into the `Next Sprint Planning` milestone to handle at a later time. The goal here will be to fill in the gaps or clarify our documentation, so that customers can be successful by using the guidance provided in the documentation.

Author: We do prioritize documentation work similar to feature work or bug fixes, so I'm not sure where this confusion is coming from. At least that's my intent to document here. Do you have a suggestion about how to better phrase that?

## Milestone Planning process

> Our milestones are usually a month old.

Reviewer: ```suggestion
Our milestones are usually a month long.
```

Before each milestone we have one or more planning meetings, where we look through all the accumulated issues in the `Next Sprint Planning` milestone and choose the most important and impactful ones to handle during the next milestone. This will be a mixture of feature requests, bug fixes, documentation issues as well as some investigations.

Note, that we will investigate only issues, which have accumulated more than certain number of upvotes and/or comments, which will indicate that there is some wider impact associated with it.

> Investigations, which accumulate no more than 5 upvotes and/or comments by different users during a period of a month will be automatically closed. The reason is that the impact is very scoped and potentially something is wrong in the user code.

Reviewer: Are we ready to pin this down to 5?  That seems like a lot based on the amount of signal VS Feedback issues tend to get.

Author: I think we can start somewhere (I've somewhat randomly chosen that number) and then adjust as we go

Reviewer: I was suggesting just leaving it vague, saying something like: "We may not investigate issues which haven't received many votes/comments."

Reviewer 2: This is by far the most controversial part of this doc and deserves expanding on.

Postponing all investigations to the next milestone seems at odds with our user engagement goals. Many people stop monitoring an issue for a response after the first few days. We should at least be immediately referring them to other resources: 
- Debugging docs
- Community forums / Stack overflow
- MS Support

There is a light weight form of investigation where we direct them what diagnostics to collect, glance through their logs, and check for common issues. This kind of community engagement needs a timely process, it should not be postponed. We're currently using the "Needs: Author Feedback" and "Needs: Attention" labels for these and we review them several times per week in triage. They only get put into a milestone once an issue is identified or the users investigation stalls and we decide it's worth developer resources to set up and debug.

Reviewer 3: I completely agree. Fast engagement can save a lot investigation later by helping bug reporters provide more details, a better repro, etc...

> Investigations, which accumulate no more than 5 upvotes and/or comments by different users during a period of a month will be automatically closed. The reason is that the impact is very scoped and potentially something is wrong in the user code.

I understand the desire to auto-close issues to reduce our workload, but I think this would allow too many things to fall under the radar. I support @Tratcher's suggestion of using the "Needs: Author feedback" as the primary mechanism for auto-closing investigations.

If it really is something wrong in the user code or environment, we can usually figure that out by asking for a simplified repro (and not getting one). Oftentimes, bug submitters will report back with what they did wrong which allows us to learn common pitfalls and gives us opportunity to try to improve or docs or APIs. Often times bug submitters will go ahead and submit a PR to improve the docs themselves.

Author: The reality is that currently we get way too many issues which have enough information to investigate - and if we try to handle all those, we will probably reduce the feature delivery velocity by at least 80%.
As an example, there are many issues which will require a lot of time to investigate yet, they are issues which only one person would hit - a corner case situation.
The model I propose here is to allow us to filter out those "non-critical" investigations and focus only on things with bigger impact.

@Tratcher , to address the engagement concern, we're going to refer to this document every time we will be moving issues to the `next sprint planning` or `backlog`, in addition to providing some note about expectations regarding when the issue will be looked at.

We do already use the `Need author feedback` label quite intensively, yet we are where we are.

@halter73, we don't neglect issues ever - we do look through all of these every day. And we do notice and cover gaps which can help avoid common pitfalls as a result of this. Here is the latest example: https://github.com/dotnet/aspnetcore/issues/22425, where the customer is again facing something which seems to be an issue on their end. As you can see we've filed a separate issue to provide an FAQ for others hitting the same in the future.
One thing this document doesn't go into detail much, is the issues where we actually answer customer questions, when these are quite straight forward. And usually, we link to documentation in these scenarios.

Reviewer 2: The `Need author feedback` process should be explicitly described in this document as well.

Reviewer 2: Let me describe this another way. The doc is missing a phase prior to investigations.

Information Gathering
In this phase we instruct the user on how to collect the appropriate diagnostics and see if they are able to address the issue with that additional information. We'll try to respond quickly to such issues (days? one week?). Issues in this phase may be closed automatically if we do not receive timely responses, they often do not provide enough information for us to investigate further. (e.g. `Needs Author Feedback`). If a user has collected all of the relevant diagnostics and the issue is still not apparent, then we will consider it for investigation by the team.

> Investigations, which accumulate no more than 5 upvotes and/or comments by different users during a period of a month will be automatically closed. The reason is that the impact is very scoped and potentially something is wrong in the user code.

Author: ```suggestion
"We may not investigate issues which haven't received many votes/comments and choose to close these. The reason is that the impact is very scoped and potentially something is wrong in the user code.
```

Managing a popular GitHub repo with a small team is not an easy task. It requires a good balance between creating new features while handling many investigations and bug fixes associated with existing ones.

During the last couple of years the amount of incoming issues has been constantly growing. While this is a sign of a healthy framework and ecosystem surrounding it, it's becoming harder to react to all those issues.

> To be able to keep up with ever-evolving expectations, we're introducing a set of rules to help us better handle the incoming issues going forward.

Author: ```suggestion
To be able to keep up with ever-evolving expectations, we're introducing a set of rules to help us better handle the incoming issues going forward.

**Note:** Customers that need help with **urgent investigations** should contact [Microsoft Support](https://dotnet.microsoft.com/platform/support).
```

## Goals
The goals of these rules are listed below in priority order:
- Make it easy to make triage decisions for each issue filed in this repository
- Be able to easily prioritize issues for each milestone
- Set proper expectations with customers regarding how issues are going to be handled

## Triage Process
The feature teams should be able look through every single issue filed against this repository and be able to make a quick call regarding the nature of the issue.
We will first categorize the issues and further handle these depending on the category the issue is in. The subsections below reprsent these categories and the rules we apply for them during our triage meeting.

### Information Gathering
In this phase we instruct the user on how to collect the appropriate diagnostics and see if they are able to address the issue with that additional information.  When we need user input we will mark the issue with `Needs: Author Feedback` label. Issues in this phase may be closed automatically if we do not receive timely responses, they often do not provide enough information for us to investigate further.

> We'll try to respond quickly to such issues (within days). If a user has collected all of the relevant diagnostics and the issue is still not apparent, then we will consider it for investigation by the team.

Reviewer 4: Link to `Investigations` label below

If we think the feature request is not aligned with our goals, we may choose to close it immediately.
In some situations, however, we may choose to collect more feedback before acting on the issue. In these situations we will move the issue in the `Backlog` so that we can review it during the [release planning](#release-planning).

### Bug reports
If it's immediately clear, that the issue is related to a bug in the framework, we will apply the `bug` label to the issue.

At this point, we will try to make a call regarding it's impact and severity. If the issue is critical, we may choose to include it in our current milestone for immediate handling or potentially patching.
If the bug is relatively high impact, we will move the issue into the `Next Sprint Planning` milestone to review during our [next sprint planning](#milestone-planning) meeting.
If the impact is unclear or the is a very corner case scenario, we may move it to the `Backlog` milestone to further evaluate the impact by reviewing customer upvotes / comments at a later time.

### Investigations

> In many situations it's not immediately clear whether a specific issue repoted is a bug or not. To be certain, the team will need to spend time to investigate it before making a call regarding the faith of the issue. In these situations we will apply the `investigate` label to the issue.

Reviewer 4: ```suggestion
In many situations it's not immediately clear whether a specific issue reported is a bug or not. To be certain, the team will need to spend time to investigate it before making a call regarding the faith of the issue. In these situations we will apply the `investigate` label to the issue.
```

In many situations, these issues turn out to be a result of some kind of misconfiguration in the user code.
In some rare situations, however, these turn out to be caused by very impactful issues. So we will make a call during the triage whether we need to immediately investigate certain issues or not.

> If not, we will move the investigation to the `Next Sprint Planning` to review during the [next sprint planning meeting](#milestone-planning).

Reviewer 2: Customers that need help with urgent investigations should contact Microsoft Support (link).

Author: Added a note here on [line 6](https://github.com/dotnet/aspnetcore/pull/22393/files#diff-42b7e1e1a33ad3f1147d9d88618f0284R6]

### Documentation requests
Some issues turn out to indicate user confusion around how to configure different aspects of the framework.
When we determine such isssues, we will mark these with the `Docs` label and move them into the `Next Sprint Planning` milestone to handle at a later time. The goal here will be to fill in the gaps or clarify our documentation, so that customers can be successful by using the guidance provided in the documentation.
If we identify a documentation issue which too many customers are having trouble with, we may choose to address that in current milestone.

## Milestone Planning
Our milestones are usually a month long.
Before each milestone we have one or more planning meetings, where we look through all the accumulated issues in the `Next Sprint Planning` milestone and choose the most important and impactful ones to handle during the next milestone. This will be a mixture of feature requests, bug fixes, documentation issues as well as some investigations.

Note, that we will investigate only issues, which have accumulated more than certain number of upvotes and/or comments, which will indicate that there is some wider impact associated with it.

> "We may not investigate issues which haven't received many votes/comments and choose to close these. The reason is that the impact is very scoped and potentially something is wrong in the user code.

Reviewer 2: ```suggestion
We may not investigate issues which haven't received many votes/comments and choose to close these. The reason is that the impact is very scoped and potentially something is wrong in the user code.
```

> We may not investigate issues which haven't received many votes/comments and choose to close these. The reason is that the impact is very scoped and potentially something is wrong in the user code.

Reviewer 4: ```suggestion
We may not investigate issues which haven't received many votes/comments and choose to close these. The reason is that the impact is very scoped and potentially something is wrong in the user code. Consider asking these questions on StackOverflow.
```

For some feature requests and bug reports, depending on the user involvement, we may choose to move these to the backlog at this point. What this means, is that they will not be looked at again up until the next major release planning.

## Release Planning
Once we approach to the end of the release cylce (.NET Core 3, .NET 5) we will look through the accumulated issues in the `Backlog` milestone. This is a long process as the amount of issues accumulated in this milestone is quite large.

We will try to prioritize issues with most community requests / upvotes assuming these are aligned with our goals.
Issues, which we will think are candidates for the upcoming release, will be moved to the `Next Sprint Planning` milestone.

## Process Visualization

> If you're a type of person who likes to see things visualized, the following diagram summarizes the processes detailed above:

Reviewer 2: ```suggestion
The following diagram summarizes the processes detailed above:
```

Reviewer 2: The chart is missing an arrow from Release Planning -> Next Sprint Planning, right now it shows everything getting closed 😁.

> ![image](https://user-images.githubusercontent.com/34246760/83341925-a03ae180-a29d-11ea-82db-e215f4860c19.png)

Author: ```suggestion
![image](https://user-images.githubusercontent.com/34246760/84076162-23d58c00-a98a-11ea-909d-6d17b91d4f84.png)
```

## Goals
The goals of these rules are listed below in priority order:
- Make it easy to make triage decisions for each issue filed in this repository
- Be able to easily prioritize issues for each milestone
- Set proper expectations with customers regarding how issues are going to be handled

## Triage Process
The feature teams should be able look through every single issue filed against this repository and be able to make a quick call regarding the nature of the issue.
We will first categorize the issues and further handle these depending on the category the issue is in. The subsections below reprsent these categories and the rules we apply for them during our triage meeting.

### Information Gathering
In this phase we instruct the user on how to collect the appropriate diagnostics and see if they are able to address the issue with that additional information.  When we need user input we will mark the issue with `Needs: Author Feedback` label. Issues in this phase may be closed automatically if we do not receive timely responses, they often do not provide enough information for us to investigate further.

> We'll try to respond quickly to such issues (within days). If a user has collected all of the relevant diagnostics and the issue is still not apparent, then we will consider it for investigation by the team.

Author: ```suggestion
We'll try to respond quickly to such issues (within days). If a user has collected all of the relevant diagnostics and the issue is still not apparent, then we will consider it for further [investigation](#investigations) by the team.
```
